src/mesh/parallel.jl

```diff
@@ -0,0 +1,114 @@
+# Partition mesh using a static domain decomposition algorithm based on leaf cell count alone
+function partition!(mesh)
+  # Determine number of leaf cells per rank
+  leaves = leaf_cells(mesh.tree)
+  @assert length(leaves) > n_mpi_ranks()
+  n_leaves_per_rank = OffsetArray(fill(div(length(leaves), n_mpi_ranks()), n_mpi_ranks()),
+                                  0:(n_mpi_ranks() - 1))
+  for d in 0:(rem(length(leaves), n_mpi_ranks()) - 1)
+    n_leaves_per_rank[d] += 1
+  end
+  @assert sum(n_leaves_per_rank) == length(leaves)
+
+  # Assign MPI ranks to all cells such that all ancestors of each cell - if not yet assigned to a
+  # rank - belong to the same rank
+  mesh.first_cell_by_rank = similar(n_leaves_per_rank)
+  mesh.n_cells_by_rank = similar(n_leaves_per_rank)
+
+  leaf_count = 0
+  last_id = leaves[n_leaves_per_rank[0]]
+  mesh.first_cell_by_rank[0] = 1
+  mesh.n_cells_by_rank[0] = last_id
+  mesh.tree.mpi_ranks[1:last_id] .= 0
+  for d in 1:(length(n_leaves_per_rank)-1)
+    leaf_count += n_leaves_per_rank[d-1]
+    last_id = leaves[leaf_count + n_leaves_per_rank[d]]
+    mesh.first_cell_by_rank[d] = mesh.first_cell_by_rank[d-1] + mesh.n_cells_by_rank[d-1]
+    mesh.n_cells_by_rank[d] = last_id - mesh.first_cell_by_rank[d] + 1
+    mesh.tree.mpi_ranks[mesh.first_cell_by_rank[d]:last_id] .= d
+  end
+
+  @assert all(x->x >= 0, mesh.tree.mpi_ranks[1:length(mesh.tree)])
+  @assert sum(mesh.n_cells_by_rank) == length(mesh.tree)
+
+  return nothing
+end
+
+
+function load_mesh(restart_filename, mpi_parallel::Val{true})
+  # Get number of spatial dimensions
+  ndims_ = parameter("ndims")
+
+  # Get maximum number of cells that should be supported
+  n_cells_max = parameter("n_cells_max")
+
+  # Create mesh
+  @timeit timer() "creation" mesh = TreeMesh(ParallelTree{ndims_}, n_cells_max)
+
+  # Determine mesh filename
+  if is_mpi_root()
+    filename = get_restart_mesh_filename(restart_filename)
+    buffer = Vector{UInt8}(filename)
+    MPI.Bcast!(Ref(length(buffer)), mpi_root(), mpi_comm())
+    MPI.Bcast!(buffer, mpi_root(), mpi_comm())
+  else # non-root ranks
+    count = MPI.Bcast!(Ref(0), mpi_root(), mpi_comm())
+    buffer = Vector{UInt8}(undef, count[])
+    MPI.Bcast!(buffer, mpi_root(), mpi_comm())
+    filename = String(buffer)
+  end
```

Comment:
I'm just thinking whether it might be nice to reduce the amount of code duplication. For example, this logic could be hidden inside `get_restart_mesh_filename` to make this part look more similar to `load_mesh(restart_filename, mpi_parallel::Val{false})`.

Reply:
Fixed in 786cc74.